### PR TITLE
Adds validation to allow only flint queries and sql SELECT queries to security lake type datasource

### LINKED ISSUE #2907: Limit Spark SQL queries to SELECT  + FLINT commands when Lake Formation is enabled
When using Lake Formation, OpenSearch should limit the Spark SQL queries to only SELECT to limit the capability of users to escalate privilege beyond what Lake Formation actually grants the user.

This proposal includes 2 main pieces:
1. Define a more limited ANTLR grammar that restricts SQL grammar to only SELECT, COVERING INDEX, and MATERIALIZED VIEW
2. If Lake Formation is enabled, validate that SQL is restricted to that grammar.

### PR DESCRIPTION
**Pull request by eirsep**
### Description
Adds validation to allow only flint queries and sql select queries to security lake type datasource

### Related Issues
Resolves #2907

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/sql/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by eirsep** [comment]
@opensearch-project/sql team, 

2 points for consideration before I can mark PR as ready
1. I am considering adding a configuration file to add validator methods for combination of datasource type and queries and register them to different datasources via config.
2. What's the right way to determine if sql query is a SELECT query? Checking if query string starts with SELECT keyword seems a bit juvenile.

File: async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java [hunk 1]
```diff
@@ -77,32 +83,73 @@ public static boolean isFlintExtensionQuery(String sqlQuery) {
...
+      sqlBaseParser.addErrorListener(new SyntaxAnalysisErrorListener());
+      SqlBaseValidatorVisitor sqlParserBaseVisitor = getSparkSqlValidatorVisitor(datasource);
+      StatementContext statement = sqlBaseParser.statement();
+      sqlParserBaseVisitor.visit(statement);
+      return sqlParserBaseVisitor.getValidationErrors();
+    } catch (SyntaxCheckException e) {
+      logger.error(
+          String.format(
+              "Failed to parse sql statement context while validating sql query %s", sqlQuery),
+          e);
```
Comment (ykmr1224) [inline async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java:100]: Why do we add this error log?

File: async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java [hunk 2]
```diff
@@ -77,32 +83,73 @@ public static boolean isFlintExtensionQuery(String sqlQuery) {
...
     try {
-      SqlBaseParser.StatementContext statement = sqlBaseParser.statement();
-      sparkSqlValidatorVisitor.visit(statement);
-      return sparkSqlValidatorVisitor.getValidationErrors();
-    } catch (SyntaxCheckException syntaxCheckException) {
+      SqlBaseParser sqlBaseParser =
+          new SqlBaseParser(
+              new CommonTokenStream(new SqlBaseLexer(new CaseInsensitiveCharStream(sqlQuery))));
+      sqlBaseParser.addErrorListener(new SyntaxAnalysisErrorListener());
+      SqlBaseValidatorVisitor sqlParserBaseVisitor = getSparkSqlValidatorVisitor(datasource);
```
Comment (ykmr1224) [inline async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java:92]: Why do we move into the try block? As we don't expect SyntaxCheckException from these, we should put outside.

File: async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java [hunk 3]
```diff
@@ -38,13 +43,14 @@
  */
 @UtilityClass
 public class SQLQueryUtils {
+  private static final Logger logger = LogManager.getLogger(SQLQueryUtils.class);
```
Comment (ykmr1224) [inline async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java:46]: nit: We prefer `LOG` instead of `logger`, though I see some variation in this repository...

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java [hunk 4]
```diff
@@ -404,15 +407,40 @@ void testAutoRefresh() {
 
   @Test
   void testValidateSparkSqlQuery_ValidQuery() {
-    String validQuery = "SELECT * FROM users WHERE age > 18";
-    List<String> errors = SQLQueryUtils.validateSparkSqlQuery(validQuery);
+    DataSource dataSource = Mockito.mock(DataSource.class);
```
Comment (ykmr1224) [inline async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java:410]: Can you use `@Mock` annotation instead?

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java [hunk 5]
```diff
@@ -404,15 +407,40 @@ void testAutoRefresh() {
 
   @Test
   void testValidateSparkSqlQuery_ValidQuery() {
```
Comment (ykmr1224) [inline async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java:409]: nit: Please use blank line to separate Arrange/Act/Assert section for readability.

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java [hunk 6]
```diff
@@ -404,15 +407,40 @@ void testAutoRefresh() {
...
-    List<String> errors = SQLQueryUtils.validateSparkSqlQuery(validQuery);
+    DataSource dataSource = Mockito.mock(DataSource.class);
+    Mockito.when(dataSource.getConnectorType()).thenReturn(DataSourceType.PROMETHEUS);
+    String validQuery = "DELETE FROM Customers WHERE CustomerName='Alfreds Futterkiste'";
+    List<String> errors = SQLQueryUtils.validateSparkSqlQuery(dataSource, validQuery);
     assertTrue(errors.isEmpty(), "Valid query should not produce any errors");
   }
 
+  @Test
+  void testValidateSparkSqlQuery_SelectQuery_DataSourceSecurityLake() {
```
Comment (ykmr1224) [inline async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java:418]: I think we want to have more test cases for positive and negative cases, such as `SELECT` query is included as part of the `CREATE` query, etc. To avoid redundancy, you'd want to extract methods like `verifyValidSecurityLakeQuery`.

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java [hunk 7]
```diff
@@ -404,15 +407,40 @@ void testAutoRefresh() {
...
+    Mockito.when(dataSource.getConnectorType()).thenReturn(DataSourceType.SECURITY_LAKE);
+    String validQuery = "SELECT * FROM users WHERE age > 18";
+    List<String> errors = SQLQueryUtils.validateSparkSqlQuery(dataSource, validQuery);
+    assertTrue(errors.isEmpty(), "Valid query should not produce any errors ");
+  }
+
+  @Test
+  void testValidateSparkSqlQuery_SelectQuery_DataSourceSecurityLake_ValidationFails() {
+    DataSource dataSource = Mockito.mock(DataSource.class);
+    Mockito.when(dataSource.getConnectorType()).thenReturn(DataSourceType.SECURITY_LAKE);
```
Comment (ykmr1224) [inline async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java:429]: nit: Can we static import?

File: async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java [hunk 8]
```diff
@@ -77,32 +83,73 @@ public static boolean isFlintExtensionQuery(String sqlQuery) {
...
+      return new SparkSqlSecurityLakeValidatorVisitor();
+    } else {
+      return new SparkSqlValidatorVisitor();
+    }
+  }
 
-    @Getter private final List<String> validationErrors = new ArrayList<>();
+  /**
+   * A generic validator impl for Spark Sql Queries that supports accumulating validation errors on
+   * visiting sql statement
```
Comment (ykmr1224) [inline async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java:117]: nit: this sounds a bit weird to me. should it say this is a base class instead of generic validator impl?

File: core/src/main/java/org/opensearch/sql/datasource/model/DataSourceType.java [hunk 9]
```diff
@@ -56,4 +57,17 @@ public static DataSourceType fromString(String name) {
...
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    DataSourceType that = (DataSourceType) o;
+    return Objects.equals(name, that.name);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(name);
+  }
```
Comment (ykmr1224) [inline core/src/main/java/org/opensearch/sql/datasource/model/DataSourceType.java:72]: Can we use `@EqualsAndHashCode` annotation instead?

If we need our own implementation, we should add unit tests.

File: async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java (hunk 1 above)
Comment (eirsep) [reply async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java:100]: for better debuggability if there is a syntax error?

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java (hunk 4 above)
Comment (eirsep) [reply async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java:410]: isn't @Mock only used for class level variables?
If yes, is it necessary to declare this variable at class level? I am using `dataSource` variable only in 2-3 methods.

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java (hunk 6 above)
Comment (eirsep) [reply async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java:418]: added

File: core/src/main/java/org/opensearch/sql/datasource/model/DataSourceType.java (hunk 9 above)
Comment (eirsep) [reply core/src/main/java/org/opensearch/sql/datasource/model/DataSourceType.java]: changed to use `@EqualsAndHashCode` annotation

File: async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java (hunk 8 above)
Comment (eirsep) [reply async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java]: fixed

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java (hunk 7 above)
Comment (eirsep) [reply async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java]: changed to static import

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java (hunk 5 above)
Comment (eirsep) [reply async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java:410]: done

File: async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java (hunk 2 above)
Comment (eirsep) [reply async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java]: reverted

File: async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java (hunk 1 above)
Comment (ykmr1224) [reply async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java:100]: As normal flow go through here, the log level should be `info` or `debug`.

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java (hunk 4 above)
Comment (ykmr1224) [reply async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java]: You can define dataSource with `@Mock` annotation once, and can use it in multiple test cases. MockitoExtension will initialize for each tests.

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java (hunk 6 above)
Comment (ykmr1224) [reply async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java:421]: As I mentioned in the above comment, can we extract method and reduce redundancy?

Comment (eirsep) [reply async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java:421]: done

File: async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java (hunk 4 above)
Comment (eirsep) [reply async-query-core/src/test/java/org/opensearch/sql/spark/utils/SQLQueryUtilsTest.java]: done

File: async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java (hunk 1 above)
Comment (eirsep) [reply async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java:100]: I am not sure if this is an info or debug level log
Imagine a bad sql query we are debugging in a log deep dive. It would be easier to trace it correctly if we throw error logs and user can also keep an eye on them.
this is an `error` level log because we have caught an exception from a user input or from another system. I strongly feel this should be retained as `error` level log

Comment (ykmr1224) [reply async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java:100]: I thought this skips SyntaxCheckException intentionally. @vamsi-amazon any thoughts?

**Review by ykmr1224 — DISMISSED** [review]

File: async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java (hunk 1 above)
Comment (vamsi-amazon) [reply async-query-core/src/main/java/org/opensearch/sql/spark/utils/SQLQueryUtils.java:100]: Yeah, I was little apprehensive to add error logs because..sometimes g4 files are not updated to the latest one from spark repo and this can result in syntax errors for valid queries. So, in case of syntax exceptions, I have decided to forward the query to spark in case of validation errors.

Now since we are clear on which spark version we are going to support, we should take g4 files from that branch and throw validation error in case of exception. what do you think @ykmr1224 ?

**Review by vamsi-amazon — COMMENTED** [review]
LGTM. If possible, can we add more types of queries in SQLQueryUtils test.

**Review by vamsi-amazon — APPROVED** [review]

**Review by ykmr1224 — APPROVED** [review]
